script/test-in-container.sh

```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+#
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+testcase=$1
+set -e pipefail
+
+[ -n "$testcase" ] || (echo No testcase provided; exit 1) >&2
+[ -f "$testcase" ] || (echo Cannot find file "$testcase"; exit 1 ) >&2
+
+thisdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+circlecidir="$thisdir"/../.circleci
+[ -d "$circlecidir" ] || circlecidir="$thisdir"/../../../.circleci
+[ -d "$circlecidir" ] || (echo Cannot find .circleci directory; exit 1 ) >&2
+basename=$(basename "$testcase")
+containername="localtest.$basename"
+
+# shellcheck source=/dev/null
+source "$thisdir/osht.sh"
+[[ ${BASH_SOURCE[0]} != *privileged.sh ]] || SKIP test "${PRIVILEGED_TESTS}" != 1 # PRIVILEGED_TESTS is not set to 1
+SKIP test "$(docker info 2>&1)" == "" # Docker doesn't seem to be running
+PLAN 1
+
+# let's use `docker build` here to untilize docker cache
+( 
+# shellcheck disable=SC2046 disable=SC2005
+echo "FROM registry.opensuse.org/devel/openqa/ci/containers/base:latest
+RUN sudo zypper ar -f -p 80 https://download.opensuse.org/repositories/devel:/openQA:/Leap:/15.1/openSUSE_Leap_15.1 openQA
+RUN sudo zypper ar -f -p 81 http://download.opensuse.org/repositories/devel:openQA/openSUSE_Leap_15.1 devel
```

Comment:
you have similar code covered already in multiple helper scripts. Please DRY. I can see you copy-pasted, because you used "https" for the first, "http" for the second in all cases ;)

Reply:
DRY easily applies inside one system with defined or obvious layers, and introducing coupling between circleci scripts, docker containers and test scripts doesn't bring any real benefits.
At the moment I don't see any big advantage of replacing flat files with some hierarchy and e.g. sourcing scripts from inside heredoc which generates Dockerfile .